src/membership/members.rs

```diff
@@ -485,4 +557,145 @@ impl<T: Trait> Module<T> {
+        // Other possible checks:
+        // How long the member has been registered
+        // Minimum balance
+        // EntryMethod
+
+        Ok(())
+    }
+
+    pub fn register_role_on_member(
+        controller_account: &T::AccountId,
+        role: Role,
+        actor_id: ActorId,
+    ) -> Result<(), &'static str> {
+        let member_id = Self::ensure_is_member_controller_account(controller_account)?;
+
+        // policy check
+        Self::can_register_role_on_member(controller_account, role)?;
+
+        // guard against duplicate ActorInRole
+        let actor_in_role = ActorInRole {
+            role_id: role.into(),
+            actor_id,
+        };
+        ensure!(
+            !<MembershipIdByActorInRole<T>>::exists(&actor_in_role),
+            "role actor already exists"
+        );
+
+        let mut profile = Self::ensure_profile(member_id)?;
+        let mut new_ids = vec![actor_id];
+
+        if let Some(current_ids) = profile.roles.get_mut(&role.into()) {
+            current_ids.append(&mut new_ids);
+        } else {
+            profile.roles.insert(role.into(), new_ids);
+        }
+        <MemberProfile<T>>::insert(member_id, profile);
+        <MembershipIdByActorInRole<T>>::insert(
+            ActorInRole {
+                role_id: role.into(),
+                actor_id,
+            },
+            member_id,
+        );
+        Self::deposit_event(RawEvent::MemberRegisteredRole(
+            member_id,
+            role.into(),
+            actor_id,
+        ));
+        Ok(())
+    }
+
+    pub fn can_unregister_role_on_member(
+        controller_account: &T::AccountId,
+        role: Role,
+        actor_id: ActorId,
+    ) -> Result<(), &'static str> {
+        let actor_in_role = ActorInRole {
+            role_id: role.into(),
+            actor_id,
+        };
+        ensure!(
+            <MembershipIdByActorInRole<T>>::exists(&actor_in_role),
+            "role actor not found"
+        );
+
+        let member_id = Self::ensure_is_member_controller_account(controller_account)?;
+        ensure!(
+            <MembershipIdByActorInRole<T>>::get(&actor_in_role) == member_id,
+            "role actor not for member"
+        );
+        Ok(())
+    }
+
+    pub fn unregister_role_on_member(
+        controller_account: &T::AccountId,
+        role: Role,
+        actor_id: ActorId,
+    ) -> Result<(), &'static str> {
+        Self::can_unregister_role_on_member(controller_account, role, actor_id)?;
+
+        let member_id = Self::ensure_is_member_controller_account(controller_account)?;
+        let mut profile = Self::ensure_profile(member_id)?;
+
+        if let Some(current_ids) = profile.roles.get_mut(&role.into()) {
+            //current_ids.remove_item(&actor_id); // unstable nightly feature!
+            current_ids.retain(|id| *id != actor_id);
+            <MemberProfile<T>>::insert(member_id, profile);
+        }
+
+        <MembershipIdByActorInRole<T>>::remove(ActorInRole {
```

Comment:
Perhaps we do not remove? In particular, if we want to hold a history of prior roles held by a member? just mark as expired?